app/controllers/course/controller.rb

```diff
@@ -16,6 +16,17 @@ def sidebar_items(type: nil)
     sidebar_items_of_type(type).sort_by { |item| weights_hash[item[:key]] || item[:weight] }
   end
 
+  # Fetches the first popup UserNotification for the current course and returns JSON data for the
+  # frontend to display it.
+  def next_popup_notification
+    notification =
+      current_user.notifications.popup.ordered_by_updated_at.unread.includes(activity: :object).
+      drop_while { |popup| popup.activity.object.course_id != current_course.id }.first
```

Comment:
This implicitly assumes all objects have a course_id attached to them, which might not necessarily be true? The next best alternative I can think of  is to include this in a scope required by all objects of notifications (something like `from_user`) - not sure if there are other ways of solving this?

Reply:
I think it is a bit premature to add a full-blown `acts_as_activity`-like framework so that we can do something like `from_user`, but it might be a good idea to consider if we have new popup types.

For now, I've changed `.course_id` to a `&.course` so that is it more generic. There is no extra cost for eager loading `course` because it is cached. New course popup activity objects can define `#course` even if they don't have `#course_id`.